baremetal/machine_manager_test.go

```diff
@@ -0,0 +1,1423 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+const (
+	testImageURL           = "http://172.22.0.1/images/rhcos-ootpa-latest.qcow2"
+	testImageChecksumURL   = "http://172.22.0.1/images/rhcos-ootpa-latest.qcow2.md5sum"
+	testUserDataSecretName = "worker-user-data"
+)
+
+func TestChooseHost(t *testing.T) {
+	scheme := runtime.NewScheme()
+	err := bmoapis.AddToScheme(scheme)
+
+	if err != nil {
+		log.Printf("AddToScheme failed: %v", err)
+	}
+
+	host1 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host1",
+			Namespace: "myns",
+		},
+		Spec: bmh.BareMetalHostSpec{
+			ConsumerRef: &corev1.ObjectReference{
+				Name:       "someothermachine",
+				Namespace:  "myns",
+				Kind:       "Machine",
+				APIVersion: capi.GroupVersion.String(),
+			},
+		},
+	}
+	host2 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host2",
+			Namespace: "myns",
+		},
+	}
+	host3 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host3",
+			Namespace: "myns",
+		},
+		Spec: bmh.BareMetalHostSpec{
+			ConsumerRef: &corev1.ObjectReference{
+				Name:       "machine1",
+				Namespace:  "myns",
+				Kind:       "Machine",
+				APIVersion: capi.GroupVersion.String(),
+			},
+		},
+	}
+	host4 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host4",
+			Namespace: "someotherns",
+		},
+	}
+	discoveredHost := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "discoveredHost",
+			Namespace: "myns",
+		},
+		Status: bmh.BareMetalHostStatus{
+			ErrorMessage: "this host is discovered and not usable",
+		},
+	}
+	hostWithLabel := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "hostWithLabel",
+			Namespace: "myns",
+			Labels:    map[string]string{"key1": "value1"},
+		},
+	}
+
+	config, infrastructureRef := newConfig(t, "", map[string]string{}, []capbm.HostSelectorRequirement{})
+	config2, infrastructureRef2 := newConfig(t, "", map[string]string{"key1": "value1"}, []capbm.HostSelectorRequirement{})
+	config3, infrastructureRef3 := newConfig(t, "", map[string]string{"boguskey": "value"}, []capbm.HostSelectorRequirement{})
+	config4, infrastructureRef4 := newConfig(t, "", map[string]string{},
+		[]capbm.HostSelectorRequirement{
+			capbm.HostSelectorRequirement{
+				Key:      "key1",
+				Operator: "in",
+				Values:   []string{"abc", "value1", "123"},
+			},
+		})
+	config5, infrastructureRef5 := newConfig(t, "", map[string]string{},
+		[]capbm.HostSelectorRequirement{
+			capbm.HostSelectorRequirement{
+				Key:      "key1",
+				Operator: "pancakes",
+				Values:   []string{"abc", "value1", "123"},
+			},
+		})
+
+	testCases := map[string]struct {
+		// input
+		Machine   capi.Machine
+		Hosts     []runtime.Object
+		BMMachine *capbm.BareMetalMachine
+		//output
+		ExpectedHostName string
+	}{
+		"1: Pick host2 which lacks ConsumerRef": {
```

Comment:
please remove these numbers from TC name. It becomes really difficult to add anything in the middle. numbers doesnt matter becasue all tests should be independent and can be run in shuffled order.

Reply:
I would say that never add a testcase in the middle if you want to avoid extra trouble. Same applies with data structures, function interfaces etc.

Anyways, can be removed later. Handy during work in progress, a lot quicker to find the failing case.

Reply:
numbering removed